text/0456-elasticache-l2.md

```diff
@@ -0,0 +1,375 @@
+or cache environment) within the CDK.
+
+### Why should I use this feature?
+
+Setting up AWS ElastiCache with AWS CDK minimizes work needed to provide access
+to other services and manage distributed in-memory data store or cache
+environments in the cloud.
+
+## Internal FAQ
+
+### Why are we doing this?
+
+There is a community offer to support with implementation right now.
+
+The [tracking Github issue for the module](https://github.com/aws/aws-cdk/issues/6908) is
+in the top 10 upvoted issues (+123) with [previous approaches to implement an L2](https://github.com/aws/aws-cdk/issues/8196)
+that were abandoned.
+
+Setting up a CacheCluster requires setting up separate SubnetGroups that could
+actually be fully automated away in CDK. Managing VPCs and SecurityGroups
+becomes significant easier compared to manual setup or setup with CloudFormation.
+
+### Why should we *not* do this?
+
+The API design is based on one customer implementation only, so it might be
+missing insights from other use cases.
+
+There are currently not enough resources to implement the full design, only
+part of it. Implementing the rest of the design later on might lead to
+information loss between now and then.
+
+### What is the technical solution (design) of this feature?
+
+- `IEndpoint` -- interface to manage addresses and ports
+
+  ```ts
+    interface IEndpoint {
+      address: string;
+      port: number;
+    }
+  ```
+
+- `IRedisReplicationGroup` -- interface for defined and imported ReplicationGroups
+  (cluster mode disabled)
+
+  ```ts
+  interface IRedisReplicationGroup extends
+      // Since RedisReplicationGroup will extend Resource
+      cdk.IResource,
+      // To open network conns between ElastiCache and services that should access it
+      ec2.IConnectable {
+    readonly primaryEndpoint: IEndpoint;
+    readonly readerEndpoint: IEndpoint;
+    allowReaderAccessFrom(peer: IConnectable, description?: string): void;
+    allowPrimaryAccessFrom(peer: IConnectable, description?: string): void;
+    grant(grantee: iam.IGrantable, ...actions: string[]): iam.Grant;
+    metric(metricName: string, props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+    // Some canned metrics as well like `metricCPUUtilization`
+  }
+  ```
+
+- `IRedisClusterReplicationGroup` -- interface for defined and imported ReplicationGroups
+  (cluster mode enabled)
+
+  ```ts
+  interface IRedisClusterReplicationGroup extends IRedisReplicationGroup {
+    readonly configurationEndpoint: IEndpoint;
+    allowConfigurationAccessFrom(peer: IConnectable, description?: string): void;
+  }
+  ```
+
+- `RedisReplicationGroupBase` -- abstract base ReplicationGroup class with some helper
+  props/methods
+
+  ```ts
+  enum CacheNodeType {
+    M6G_LARGE = 'cache.m6g.large',
+    // ...
+  }
+  
+  enum RedisEngineVersion {
+    VER_6_2 = '6.2',
+    // ...
+  }
+
+  interface RedisReplicationGroupBaseProps {
+    engineVersion: RedisEngineVersion,
```

Comment:
Can we default to a version similarly to what RDS does?

Reply:
Which RDS constructs do you mean? All constructs I checked when writing this RFC require `engine` to be explicitly set - and from my point of view this makes sense, as, otherwise, updating your `aws-cdk" dependency could trigger a database version upgrade, which might be not a nice surprise.

Reply:
I would argue that engine should be explicitly set. We've had issues in the past where EC changes our default version and something breaks on the end user side.

Reply:
Yep engine needs to be a required property.

Reply:
Resolving this without a change to the RFC.